Add functionality to select WordPress version to Add site form and Onboarding

bcotrim opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes https://github.com/Automattic/dotcom-forge/issues/10531

## Proposed Changes

- Added the ability to select PHP and WordPress versions during the add site and on-boarding workflows
- Ensured selected versions are passed to the site creation process
- Implemented conditional rendering of version selection UI based on the STUDIO_WP_VERSIONS feature flag
- Added automatic downloading of WordPress versions if they don't exist locally

| Add Site | Onboarding |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/7c5ae5c7-0110-4a8b-b8e3-9e6bdeda77a7) | ![image](https://github.com/user-attachments/assets/9cd341b2-017a-4800-881f-7203dc95cb68) |

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Set the STUDIO_WP_VERSIONS environment variable to true when launching the app 
```bash
STUDIO_WP_VERSIONS=true npm start
```
- Go through the on-boarding workflow (delete all sites if needed)
- Verify that PHP and WordPress version dropdowns appear in the advanced settings section
- Select different PHP and WordPress versions and complete the site creation
- Verify that the selected versions are applied to the created site
  - Check in Studio "Settings" for the website
  - Check in the site `wp-admin` for WordPress and PHP versions
- Test with the feature flag disabled to ensure the version selection UI doesn't appear
- Try selecting a WordPress version that isn't already downloaded to verify it's automatically downloaded during site creation

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

bcotrim committed:
pick wordpress and php versions

bcotrim committed:
add site php and wordpress versions

bcotrim committed:
add feature flag condition

bcotrim committed:
use wordpress version from store

bcotrim committed:
add unit tests

bcotrim committed:
fix lint

bcotrim committed:
fix lint

bcotrim committed:
fix lint

bcotrim committed:
refactor unit tests

katinthehatsite commented on src/components/site-form.tsx:
```diff
@@ -352,6 +395,50 @@ export const SiteForm = ( {
+													/>
+												</div>
+												<div className="flex flex-col gap-1.5 leading-4">
+													<label className="font-semibold">{ __( 'WordPress version' ) }</label>
```
Let's set the fields for PHP and WordPress versions to be 40px in height to be consistent with other fields in the form and the proposed design: 

<img width="263" alt="Screenshot 2025-02-28 at 10 17 13 AM" src="https://github.com/user-attachments/assets/7a60ac2b-5d6c-45ce-a90e-128324e7688a" />

katinthehatsite commented on src/components/site-form.tsx:
```diff
@@ -212,24 +259,20 @@ export const SiteForm = ( {
-	setFileForImport?: ( file: File | null ) => void;
-	onFileSelected?: ( file: File ) => void;
-	fileError?: string;
-} ) => {
```
Nice!

katinthehatsite commented on src/components/tests/add-site.test.tsx:
```diff
@@ -209,10 +264,90 @@ describe( 'AddSite', () => {
 		await user.click( screen.getByTestId( 'select-path-button' ) );
 		await user.click( screen.getByDisplayValue( 'My WordPress Website' ) );
 		await user.type( screen.getByDisplayValue( 'My WordPress Website' ), ' mutated' );
-		// screen.debug( screen.getByRole( 'dialog' ) );
```
Should we remove this?

katinthehatsite commented on src/components/tests/onboarding.test.tsx:
```diff
@@ -31,11 +43,36 @@ jest.mock( 'src/lib/get-ipc-api', () => ( {
 	} ),
 } ) );
 
+// Mock dispatch function
```
I think we can clean up these comments as they are self-explanatory

katinthehatsite commented on src/components/tests/onboarding.test.tsx:
```diff
@@ -72,4 +119,263 @@ describe( 'Onboarding Component', () => {
+			return { status: 'succeeded' };
+		} );
+
+		// Mock dispatch to return a function that resolves immediately
```
Similar comment about cleaning up comments

katinthehatsite commented:
The functionality worked well for me, I left a couple of comments for the first review. I will review in closer detail on Monday - so far looks good 😍

bcotrim committed:
set height 40px

bcotrim committed:
remove unnecessary code comments

bcotrim committed:
fix

sejas reviewed (commented):
In the Onboarding form, a scroll appears and it produces a weird space at the bottom:

<img width="891" alt="Screenshot 2025-03-03 at 16 51 18" src="https://github.com/user-attachments/assets/2ccd6230-9336-4ff3-b1ce-f283091d7bbe" />


I tested the feature by creating multiple sites using both the onboarding and regular add site forms, confirming that everything worked as expected and that the PHP and WordPress versions were correctly applied.


https://github.com/user-attachments/assets/8960de0f-9cb9-4f80-8b58-75f18ec63385

sejas commented on src/components/site-form.tsx:
```diff
@@ -352,6 +395,50 @@ export const SiteForm = ( {
 											value={ sitePath }
 											onClick={ onSelectPath }
 										/>
+										{ wpVersionsEnabled && allowVersionsChange && (
```
I don't fully understand why we need the `allowVersionsChange` prop.

bcotrim replied:
We can also remove that once the edit page component has the versions.
I added more details in TODO comments.

sejas commented on src/components/site-form.tsx:
```diff
@@ -352,6 +395,50 @@ export const SiteForm = ( {
+														value={ wpVersion }
+														options={
+															wpVersions.length > 0
+																? wpVersions.map( ( v ) => ( { label: v.name, value: v.version } ) )
```
When you merge trunk you'll be able to pass the wpVersions object without any transformation.

bcotrim replied:
Updated! Thanks for catching this one

sejas commented on src/components/site-form.tsx:
```diff
@@ -352,6 +395,50 @@ export const SiteForm = ( {
+																		},
+																  ]
+														}
+														onChange={ ( version ) => {
+															if ( setWpVersion ) {
+																setWpVersion( version );
+															}
+														} }
```
```suggestion
														onChange={ setWpVersion }
```

sejas commented on src/components/site-form.tsx:
```diff
@@ -35,6 +43,41 @@ interface SiteFormErrorProps {
 	className?: string;
 }
 
+interface SiteFormBaseProps {
+	className?: string;
+	children?: React.ReactNode;
+	siteName: string;
+	setSiteName: ( name: string ) => void;
+	sitePath?: string;
+	onSelectPath?: () => void;
+	error: string;
+	doesPathContainWordPress?: boolean;
+	isPathInputDisabled?: boolean;
+	onSubmit: ( event: FormEvent ) => void;
+	fileForImport?: File | null;
+	setFileForImport?: ( file: File | null ) => void;
+	onFileSelected?: ( file: File ) => void;
+	fileError?: string;
+}
+
+interface SiteFormWithVersionsProps extends SiteFormBaseProps {
+	allowVersionsChange: true;
+	phpVersion: SupportedPHPVersion;
+	setPhpVersion: ( version: SupportedPHPVersion ) => void;
+	wpVersion: string;
+	setWpVersion: ( version: string ) => void;
+}
+
+interface SiteFormWithoutVersionsProps extends SiteFormBaseProps {
+	allowVersionsChange?: false;
+	phpVersion?: never;
+	setPhpVersion?: never;
+	wpVersion?: never;
+	setWpVersion?: never;
+}
```
We might not need this complexity once we [remove the feature flag](https://github.com/Automattic/dotcom-forge/issues/10557). I see we use `SiteForm` in `EditSite`, but that component will be removed in favor of [`EditSiteDetails`](https://github.com/Automattic/studio/blob/bd081d77cafc0db645299bc1223e4e5827c492c3/src/modules/site-settings/edit-site-details.tsx#L13).

So we could make all those props required instead of optionals. We can do it as a separate PR once we've removed the feature flag.

https://github.com/Automattic/studio/blob/bd081d77cafc0db645299bc1223e4e5827c492c3/src/components/content-tab-settings.tsx#L50

bcotrim replied:
Sounds good!
I added a TODO comment to remove this once the edit feature is merged.

sejas commented on src/hooks/use-add-site.ts:
```diff
@@ -1,20 +1,33 @@
+	);
+	const [ wpVersion, setWpVersion ] = useState( DEFAULT_WORDPRESS_VERSION );
+
+	useEffect( () => {
+		setPhpVersion( DEFAULT_PHP_VERSION as SupportedPHPVersion );
+		setWpVersion( DEFAULT_WORDPRESS_VERSION );
+	}, [] );
```
It seems this useEffect that sets the default versions is not necessary 🤔 .

bcotrim replied:
Nice catch! I've removed it

sejas reviewed (commented):
Everything works as expected. Great work! I've left some questions and suggestions in the code.

bcotrim committed:
fix overflow issue on onboarding page

bcotrim committed:
remove unnecessary use effect

bcotrim committed:
add comment to simplify types once edit is done

bcotrim committed:
add more details in todo comments and simplify function calls

bcotrim committed:
cleanup wpversion after rebase

bcotrim commented:
Thanks for the reviews @katinthehatsite @sejas
I think I've addressed all the feedback.
Please let me know if you have any other feedback or adjustments.

bcotrim committed:
fix unit tests

katinthehatsite reviewed (approved):
The changes look good ❤️ 

One last request - let's slightly reduce the spacing between version selection and the `Add site` button:

<img width="484" alt="Screenshot 2025-03-03 at 8 03 17 PM" src="https://github.com/user-attachments/assets/86e9ee89-5599-4d30-bbd4-7952c636f4f5" />

katinthehatsite commented:
@bcotrim I approved the PR but let's wait for the release of 1.3.5 before we merge.

bcotrim committed:
fix spacing between advanced settings and add site

sejas reviewed (approved):
Thanks for the updates. It works as expected and I can create sites with different PHP and WP versions.

After updating from trunk, I noticed that the WordPress selector displays the beta versions first. I'll create a PR to prepend the `latest` version in the redux list of versions.

<img width="1012" alt="Screenshot 2025-03-05 at 13 00 29" src="https://github.com/user-attachments/assets/ba0ac1a1-619f-4760-9235-821dc1645220" />

I also noticed that when I tried to create the beta version, It didn't work. 
I've created an issue as a follow-up https://github.com/Automattic/dotcom-forge/issues/10567

A last slight improvement we can consider is scrolling down when the advance settings expand in the onboarding and add site modal. As an alternative we could use the floating "bottom" action bar, like we do in the Assistant or Preview tabs.
Now I need to click on it, and then scroll down to see the add site button.
No further actions are needed for this PR. I'll leave it to @katinthehatsite to decide if we should implement any logic or design change.

<img width="1012" alt="Screenshot 2025-03-05 at 13 02 12" src="https://github.com/user-attachments/assets/8a9b99cf-e1cd-4629-8768-f0ead6af230e" />

<img width="1012" alt="Screenshot 2025-03-05 at 13 16 01" src="https://github.com/user-attachments/assets/506eeaaa-db40-4341-a971-d740f0c05b72" />

sejas commented:
> After updating from trunk, I noticed that the WordPress selector displays the beta versions first. I'll create a PR to prepend the latest version in the redux list of versions.

- @bcotrim , to select the `latest` version I've created this PR https://github.com/Automattic/studio/pull/1010 . Let me know what you think.

katinthehatsite commented:
It looks like we might need to rebase the changes to trunk as now it shows just `latest`:

<img width="499" alt="Screenshot 2025-03-06 at 11 43 13 AM" src="https://github.com/user-attachments/assets/2ac0e7ed-9c0d-4777-97da-12fc2f54233e" />

bcotrim committed:
show latest versions

bcotrim commented:
I rebased `trunk` and made some changes to show the latest version.
Let me know if you have any other feedback or find anything strange please.

katinthehatsite commented:
@bcotrim I think we are ready to merged 👍 One thing I see that it now throws some conflicts with edit site details form so let's make sure to resolve those and go ahead.

fredrikekelund commented:
@bcotrim, it looks like this PR introduced a misalignment between the `Local path` field and the other fields in the `Add site` modal. See https://github.com/Automattic/dotcom-forge/issues/10582 for more details
